### PR TITLE
[VIRTS-3835] ensure potential links have fact

### PR DESCRIPTION
**Pull request by bleepbop**
## Description
Previously, there was an issue with adding facts to running operations using the Facts/Relationships API (facts can be tied to an operation by adding the operation_id as the `source` field value). API-supplied facts would be listed in the operation report, but would not be applied to any potential links that used them.

This required a few changes. First,  I added two new endpoints: `GET /facts/{operation_id}` and `GET /relationships/{operation_id}`. Previously, chrome was throwing the following error when supplying a body in a `GET` request:

Screen Shot 2022-03-09 at 2 10 45 PM

In this portion of the UI, we are primarily querying facts associated with an operation; these new endpoints allow us to do so without supplying a message body. The original endpoints have been kept in place in case users want to query facts by other criteria.

The other main change required was to update `getFacts()` in `templates/operations.html` to use the new `GET /api/v2/facts/{operation_id}` endpoint. Previously, this API call was not returning anything, but it now returns the facts associated with the selected operation.

## Type of change
- [X] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
Verified that running operation will use facts added via v2 Fact/Relationship API. Updated `test_knowledge.py`.

## Checklist:
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by argaudreau — APPROVED** [review]
Tested locally by running a couple of operations that collected facts, ran with no issues. LGTM for the front-end

**Comment by uruwhy** [comment]
@argaudreau @bleepbop Food for thought - would it make more sense to have the endpoints be `/facts/create` and `/facts/retrieve` (similarly /relationships/create and /relationships/retrieve)

```
POST /facts/create # create fact, previously POST /facts
POST /relationships/create # create relationship, previously POST /relationships
POST /facts/retrieve # get facts, previously GET /facts
POST /relationships/retrieve # get relationships, previously GET /relationships
```

**Comment by uruwhy** [comment]
Also there are a few sonarcloud things that can be fixed by making the endpoint paths constants

**Comment by bleepbop** [comment]
> @argaudreau @bleepbop Food for thought - would it make more sense to have the endpoints be `/facts/create` and `/facts/retrieve` (similarly /relationships/create and /relationships/retrieve)
> 
> ```
> POST /facts/create # create fact, previously POST /facts
> POST /relationships/create # create relationship, previously POST /relationships
> POST /facts/retrieve # get facts, previously GET /facts
> POST /relationships/retrieve # get relationships, previously GET /relationships
> ```

I believe this came up with a previous `facts_api.py` PR, but I ended up not making any changes to the endpoint names. From the discussion then, it seemed like we wanted the facts API endpoints to be similar to/consistent with other object endpoints (where creation is just `POST /object_name`), which is why I only renamed the `GET` requests

Maybe a better solution would be to keep these existing `GET` requests as-is, but add in a new endpoints (`GET /api/v2/facts/{source_id}` and `GET /api/v2/relationships/{source_id}`, which would return all facts/relationships associated with a `source_id`) so that the endpoint naming is consistent?

**Comment by uruwhy** [comment]
> > @argaudreau @bleepbop Food for thought - would it make more sense to have the endpoints be `/facts/create` and `/facts/retrieve` (similarly /relationships/create and /relationships/retrieve)
> > ```
> > POST /facts/create # create fact, previously POST /facts
> > POST /relationships/create # create relationship, previously POST /relationships
> > POST /facts/retrieve # get facts, previously GET /facts
> > POST /relationships/retrieve # get relationships, previously GET /relationships
> > ```
> 
> I believe this came up with a previous `facts_api.py` PR, but I ended up not making any changes to the endpoint names. From the discussion then, it seemed like we wanted the facts API endpoints to be similar to/consistent with other object endpoints (where creation is just `POST /object_name`), which is why I only renamed the `GET` requests
> 
> Maybe a better solution would be to keep these existing `GET` requests as-is, but add in a new endpoints (`GET /api/v2/facts/{source_id}` and `GET /api/v2/relationships/{source_id}`, which would return all facts/relationships associated with a `source_id`) so that the endpoint naming is consistent? This would prevent breaking changes from renaming the original endpoints and changing their HTTP methods.

I like this idea - @argaudreau any issues on your end?

File: tests/api/v2/test_knowledge.py [hunk 1]
```diff
@@ -107,6 +127,38 @@ async def test_display_relationships(knowledge_webapp, aiohttp_client):
...
+    resp = await client.get(f'/relationships/{op_id_test}', headers=headers)
+    data = await resp.json()
+    response = data['found']
+
+    assert len(response) == 1
+    assert response[0]['source']['trait'] == 'a'
+    assert response[0]['source']['value'] == '1'
+    assert response[0]['edge'] == 'gamma'
+    assert response[0]['origin'] == 'this_is_a_valid_operation_id'
+    assert response[0]['source']['source'] == 'this_is_a_valid_operation_id'
```
Comment (uruwhy) [inline tests/api/v2/test_knowledge.py:159]: can we also verify the target?

Comment (bleepbop) [reply tests/api/v2/test_knowledge.py]: should be updated now to check for target

**Review by uruwhy — APPROVED** [review]
LGTM! Nice work!

**Comment by argaudreau** [comment]
> > > @argaudreau @bleepbop Food for thought - would it make more sense to have the endpoints be `/facts/create` and `/facts/retrieve` (similarly /relationships/create and /relationships/retrieve)
> > > ```
> > > POST /facts/create # create fact, previously POST /facts
> > > POST /relationships/create # create relationship, previously POST /relationships
> > > POST /facts/retrieve # get facts, previously GET /facts
> > > POST /relationships/retrieve # get relationships, previously GET /relationships
> > > ```
> > 
> > 
> > I believe this came up with a previous `facts_api.py` PR, but I ended up not making any changes to the endpoint names. From the discussion then, it seemed like we wanted the facts API endpoints to be similar to/consistent with other object endpoints (where creation is just `POST /object_name`), which is why I only renamed the `GET` requests
> > Maybe a better solution would be to keep these existing `GET` requests as-is, but add in a new endpoints (`GET /api/v2/facts/{source_id}` and `GET /api/v2/relationships/{source_id}`, which would return all facts/relationships associated with a `source_id`) so that the endpoint naming is consistent? This would prevent breaking changes from renaming the original endpoints and changing their HTTP methods.
> 
> I like this idea - @argaudreau any issues on your end?

No issues here!

**Pull request merged**